### PR TITLE
[NVIDIA] Support BNTH input formats for the fused attention

### PR DESCRIPTION
**Pull request by kaixih**
For enhanced performance and API flexibility, we've extended the functionality of `dot_production_attention` to accommodate QKV inputs in the BNTH format.

cc. @Cjkkkk

### CODE REVIEW

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 1]
```diff
@@ -28,14 +28,26 @@
...
+from jax._src import util
 from jax._src.interpreters import batching
 from jax._src.lib import cuda_versions
+from jax._src.lib import xla_extension_version
 
 Array = jnp.ndarray
 DType = jnp.dtype
 PRNGKey = jnp.ndarray
 
+def _normalize_layout(layout_str):
```
Comment (Cjkkkk) [inline jax/_src/cudnn/fused_attention_stablehlo.py:44]: I think choosing between btnh and bnth is good enough.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 2]
```diff
@@ -100,50 +103,51 @@ def create_dot_product_attention_backend_config(batch,
...
+    dims = [
+        ((3, 3), ((0, 2), (0, 2))), # BMM1: BTNH,BSNH->BNTS
+        ((3, 1), ((0, 1), (0, 2))), # BMM2: BNTS,BSNH->BTNH
+        ((2, 1), ((0, 1), (0, 2))), # BMM1_grad_1: BNTS,BTNH->BSNH
+        ((3, 1), ((0, 1), (0, 2))), # BMM1_grad_2: BNTS,BSNH->BTNH
+        ((2, 1), ((0, 1), (0, 2))), # BMM2_grad_1: BNTS,BTNH->BSNH
+        ((3, 3), ((0, 2), (0, 2))), # BMM2_grad_2: BTNH,BSNH->BNTS
+    ]
+  keys = [
+      "bmm1_dot_dimension_numbers",
```
Comment (Cjkkkk) [inline jax/_src/cudnn/fused_attention_stablehlo.py:129]: why not just merge dim and keys into a dict

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 3]
```diff
@@ -178,42 +182,50 @@ def create_dot_product_attention_backend_config(batch,
...
-  _, kv_sqe_len, _, _ = key.shape
-  is_cross_attention = q_seq_len != kv_sqe_len
-  # check if attention pattern is supported by flash attention or fused attention
-  if q_seq_len > 512 and kv_sqe_len > 512 and head_dim in [64, 128]:
-    # check if flash attention is supported
+def check_qkv_layout(query, key, value, layout):
+  def assert_eq(a, b, c, msg):
+    assert a == b == c, msg + f' must be same: {a}, {b}, {c}'
+
+  Q_rank, K_rank, V_rank  = len(query.shape), len(key.shape), len(value.shape)
```
Comment (Cjkkkk) [inline jax/_src/cudnn/fused_attention_stablehlo.py:189]: maybe use q_rank, k_rank

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 4]
```diff
@@ -178,42 +182,50 @@ def create_dot_product_attention_backend_config(batch,
...
+    qB, qN,  _, qH = query.shape
+    kB, kN, kS, kH = key.shape
+    vB, vN, vS, vH = value.shape
+  else:
+    assert layout == 'BTNH'
+    qB,  _, qN, qH = query.shape
+    kB, kS, kN, kH = key.shape
+    vB, vS, vN, vH = value.shape
+  
+  assert_eq(qB, kB, vB, 'QKV batch')
```
Comment (Cjkkkk) [inline jax/_src/cudnn/fused_attention_stablehlo.py:207]: For this part, i think the original implementation is good enough, let's keep unrelated changes minimum.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 5]
```diff
@@ -228,61 +240,81 @@ def check_cudnn_version(is_flash_attention, is_cross_attention):
...
-def _dot_product_attention_fwd(query, key, value, bias, mask,
-  scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask):
+def _dot_product_attention_fwd(
+    query, key, value, bias, mask, scale, seed, dropout_rate, variadic_args,
+    is_flash_attention, is_causal_mask, layout):
   output, _ = _dot_product_attention_fwd_p_wrapper.bind(
-    query, key, value, bias, mask, scale=scale, seed=seed, dropout_rate=dropout_rate,
-    variadic_args=variadic_args, is_flash_attention=is_flash_attention,
-    is_causal_mask=is_causal_mask)
+      query, key, value, bias, mask, scale=scale, seed=seed,
```
Comment (Cjkkkk) [inline jax/_src/cudnn/fused_attention_stablehlo.py:247]: for format changes like this, maybe include in a separate pr so reviewers can review the changes related to layout changes easily.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 6]
```diff
@@ -604,22 +678,59 @@ def _infer_bwd_output_sharding(mesh, arg_shapes, variadic_args):
...
+      _dot_product_attention_bwd_impl, scale=scale, seed=seed,
+      dropout_rate=dropout_rate, variadic_args=variadic_args,
+      is_flash_attention=is_flash_attention, is_causal_mask=is_causal_mask,
+      layout=layout,
+  )
   return mesh, impl, out_shardings, arg_shardings
 
+# We redefine the xla.apply_primitive here since we need to set the jit option
+# static_argnames='layout' to ensure that the layout (str) is static.
+def apply_primitive(prim, *args, **params):
```
Comment (Cjkkkk) [inline jax/_src/cudnn/fused_attention_stablehlo.py:706]: why is this required?

Comment (kaixih) [reply jax/_src/cudnn/fused_attention_stablehlo.py:706]: It is mainly to specify the `static_argnames` in api.jit(). Otherwise, the compiler will complain the str is not a supported jax type.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 5 above)
Comment (kaixih) [reply jax/_src/cudnn/fused_attention_stablehlo.py:247]: When I add the `layout=layout` to the callsite, the pylint will give format violation warning. Do you mean we should ignore all format issues? I thought for every commit we should follow the google code style, like https://google.github.io/styleguide/pyguide.html.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 4 above)
Comment (kaixih) [reply jax/_src/cudnn/fused_attention_stablehlo.py:207]: To support a different layout, we need to manipulate the names of different dims. So, I make them more concise and remove duplicate code. So, I think this is related.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 2 above)
Comment (kaixih) [reply jax/_src/cudnn/fused_attention_stablehlo.py:129]: I think this is mainly for format and readability. Putting the comment besides the dims is more informative, I think.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 7]
```diff
@@ -604,22 +678,59 @@ def _infer_bwd_output_sharding(mesh, arg_shapes, variadic_args):
...
-                variadic_args=variadic_args, is_flash_attention=is_flash_attention, is_causal_mask=is_causal_mask)
+  impl = partial(
+      _dot_product_attention_bwd_impl, scale=scale, seed=seed,
+      dropout_rate=dropout_rate, variadic_args=variadic_args,
+      is_flash_attention=is_flash_attention, is_causal_mask=is_causal_mask,
+      layout=layout,
+  )
   return mesh, impl, out_shardings, arg_shardings
 
+# We redefine the xla.apply_primitive here since we need to set the jit option
```
Comment (yashk2810) [inline jax/_src/cudnn/fused_attention_stablehlo.py:704]: Please don't do this.

Comment (yashk2810) [reply jax/_src/cudnn/fused_attention_stablehlo.py:704]: we are discussing other approaches offline :)

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 1 above)
Comment (kaixih) [reply jax/_src/cudnn/fused_attention_stablehlo.py:49]: Passing the str layout makes the API more generic and scalable. But, ss you suggested, I make the internal representation as an enum of two possible values. This can also help us avoid passing str into the jit function, which can help us avoid the hassle of the intricacy of setting `static_argnum/names`.

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 8]
```diff
@@ -28,14 +29,30 @@
...
 from jax.sharding import PartitionSpec, NamedSharding
 
+from jax._src import api
+from jax._src import config
 from jax._src import dispatch
+from jax._src import lib
+from jax._src import util
 from jax._src.interpreters import batching
 from jax._src.lib import cuda_versions
+from jax._src.lib import xla_extension_version
```
Comment (Cjkkkk) [inline jax/_src/cudnn/fused_attention_stablehlo.py:39]: Remove this and other import as they are not used anymore.

**Comment by kaixih** [comment]
Just rebased the change to resolve some conflicts. Also tried to minimize the changes.

**Comment by Cjkkkk** [comment]
@kaixih Thanks for the PR, LGTM.

**Comment by kaixih** [comment]
@Cjkkkk Do you know who was the reviewer for such changes to your previous attn changes? Maybe we can ping him/her?

**Comment by Cjkkkk** [comment]
> @Cjkkkk Do you know who was the reviewer for such changes to your previous attn changes? Maybe we can ping him/her?

Peter just did the work :)

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 9]
```diff
@@ -36,6 +37,17 @@
 DType = jnp.dtype
 PRNGKey = jnp.ndarray
 
+class AttentionLayout(Enum):
+  BTNH = 0
+  BNTH = 1
+
+def _normalize_layout(layout_str):
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:44]: How about just `layout`? You can use a type annotation to document the type:

```
def _normalize_layout(layout: str) -> AttentionLayout:
  ...
```

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 10]
```diff
@@ -178,46 +191,56 @@ def create_dot_product_attention_backend_config(batch,
...
-def check_is_flash_attention(query, key, cudnn_version, has_bias, is_training):
-  batch, q_seq_len, num_heads, head_dim = query.shape
-  _, kv_seq_len, _, _ = key.shape
-
-  # check if attention pattern is supported by flash attention or fused attention
-  if q_seq_len <= 512 and kv_seq_len <= 512 and head_dim == 64 \
-    and (not is_training or q_seq_len % 64 == 0 and kv_seq_len % 64 == 0):
+def check_qkv_layout(query, key, value, layout):
+  def assert_eq(a, b, c, msg):
+    assert a == b == c, msg + f' must be same: {a}, {b}, {c}'
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:196]: I can see that the old version also used `assert` for argument validation, but since you are changing the validation logic slightly, I would recommend using `raise` instead. For example:

```
if q_rank != 4:
  raise ValueError(f"Q must have a ran of 4, got {q_rank}")
```

File: jax/_src/cudnn/fused_attention_stablehlo.py [hunk 11]
```diff
@@ -751,40 +846,51 @@ def dot_product_attention(query: Array,
...
-  b kv_seq num_heads head_dim -> K
-  b kv_seq num_heads head_dim -> V
+  """Computes dot-product attention given query (Q), key (K), and value (V).
+
+  This function serves as the core operation for applying attention
+  mechanisms as described in the paper [https://arxiv.org/abs/1706.03762].
+  Initially, it determines the attention weights by processing Q and K,
+  subsequently combining the outcomes using K. Throughout this function, we
+  utilize the following uppercase letters to represent specific parameters of
+  JTensor:
```
Comment (superbobry) [inline jax/_src/cudnn/fused_attention_stablehlo.py:858]: What's a JTensor? Perhaps you meant "array"?

**Review by superbobry — APPROVED** [review]

**Comment by Cjkkkk** [comment]
@superbobry Hey Sergei. any updates on this?

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 9 above)
Comment (kaixih) [reply jax/_src/cudnn/fused_attention_stablehlo.py]: Yes, this is better. Fixed.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 10 above)
Comment (kaixih) [reply jax/_src/cudnn/fused_attention_stablehlo.py]: Done.

File: jax/_src/cudnn/fused_attention_stablehlo.py (hunk 11 above)
Comment (kaixih) [reply jax/_src/cudnn/fused_attention_stablehlo.py]: Done.

**Comment by superbobry** [comment]
Can you squash the commits, please?

**Comment by kaixih** [comment]
> Can you squash the commits, please?

Done. Also removed the trailing spaces pointed by the failed lint tests.

**Comment by kaixih** [comment]
Now, it seems all tests pass.
@superbobry PTAL.

**Comment by Cjkkkk** [comment]
@superbobry Hi, any updates on this?
